### PR TITLE
Bot stats

### PR DESCRIPTION
**Pull request by titulebolide**
It is the follow up from #210 since @ullasbharadwaj deleted his repo

What's left to do:

- [x] Use the table.tabularize tables instead of prettytable
- [x] Take care that long coin list will not trigger Too long message errors
- [x] Use translation files for bot_stats
- [x] Fix type issue with tables (floats are not recognized as floats, hence display issues)
- [x] Choose how to display profits

### CODE REVIEW

**Comment by ullasbharadwaj** [comment]
I am sorry, deleted wrong repo, I had a different fork from long ago and instead of deleting that, I deleted the latest one :-( I am sorry.

**Comment by titulebolide** [comment]
No worries!

**Comment by titulebolide** [comment]
@all-contributors please add @ullasbharadwaj for ideas,code

**Comment by titulebolide** [comment]
@all-contributors please add @ullasbharadwaj for ideas, code

File: btb_manager_telegram/buttons.py [hunk 1]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+    message = [i18n_format("database_not_found", path=db_file_path)]
+    if not os.path.exists(db_file_path):
+        return message
+    message = ""
+    try:
+        con = sqlite3.connect(db_file_path)
+
+        cur = con.cursor()
+
+        cur.execute("select symbol from coins where enabled=1")
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:539]:
```suggestion
        cur.execute("SELECT symbol FROM coins WHERE enabled=1")
```

File: btb_manager_telegram/buttons.py [hunk 2]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+        bot_start_date = cur.fetchall()[0][0]
+
+        cur.execute("SELECT datetime FROM scout_history order by id desc limit 1")
+        bot_end_date = cur.fetchall()[0][0]
+
+        cur.execute("SELECT * FROM trade_history ")
+        lenTradeHistory = len(cur.fetchall())
+
+        cur.execute(
+            "SELECT alt_coin_id FROM trade_history where id=1 and state='COMPLETE' order by id asc limit 1"
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:555]:
```suggestion
            "SELECT alt_coin_id FROM trade_history WHERE id=1 and state='COMPLETE' ORDER BY id ASC LIMIT 1"
```

File: btb_manager_telegram/buttons.py [hunk 3]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+        )
+        firstTradeCoin = cur.fetchall()[0][0]
+
+        initialCoinID = ""
+        for i in range(1, lenTradeHistory):
+            cur.execute(
+                "SELECT alt_coin_id FROM trade_history where id='{}' and state='COMPLETE' order by id asc limit 1".format(
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:562]:
```suggestion
                "SELECT alt_coin_id FROM trade_history WHERE id='{}' and state='COMPLETE' ORDER BY id ASC LIMIT 1".format(
```

File: btb_manager_telegram/buttons.py [hunk 4]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+            coinID = cur.fetchall()
+            if len(coinID) > 0:
+                coinID = coinID[0][0]
+            else:
+                continue
+            for coin in coinList:
+                if coinID == coin[0]:
+                    initialCoinID = coinID
+                    cur.execute(
+                        "select alt_trade_amount from trade_history where alt_coin_id='{}' and state='COMPLETE' order by id asc limit 1".format(
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:575]:
```suggestion
                        "SELECT alt_trade_amount FROM trade_history WHERE alt_coin_id='{}' and state='COMPLETE' ORDER BY id ASC LIMIT 1".format(
```

File: btb_manager_telegram/buttons.py [hunk 5]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+                            initialCoinID
+                        )
+                    )
+                    initialCoinValue = cur.fetchall()[0][0]
+
+                    cur.execute(
+                        "select crypto_trade_amount from trade_history where alt_coin_id='{}' and state='COMPLETE' order by id asc limit 1".format(
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:582]:
```suggestion
                        "SELECT crypto_trade_amount FROM trade_history WHERE alt_coin_id='{}' and state='COMPLETE' ORDER BY id ASC LIMIT 1".format(
```

File: btb_manager_telegram/buttons.py [hunk 6]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+                            initialCoinID
+                        )
+                    )
+                    initialCoinFiatValue = cur.fetchall()[0][0]
+                    break
+            if initialCoinID != "":
+                break
+
+        cur.execute(
+            "select alt_coin_id from trade_history where selling=0 and state='COMPLETE' order by id desc limit 1"
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:592]:
```suggestion
            "SELECT alt_coin_id FROM trade_history WHERE selling=0 and state='COMPLETE' ORDER BY id DESC LIMIT 1"
```

File: btb_manager_telegram/buttons.py [hunk 7]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+        )
+        lastCoinID = cur.fetchall()[0][0]
+
+        cur.execute(
+            "select alt_trade_amount from trade_history where selling=0 and state='COMPLETE' order by id desc limit 1"
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:597]:
```suggestion
            "SELECT alt_trade_amount FROM trade_history WHERE selling=0 and state='COMPLETE' ORDER BY id DESC LIMIT 1"
```

File: btb_manager_telegram/buttons.py [hunk 8]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+        )
+        lastCoinValue = cur.fetchall()[0][0]
+
+        cur.execute(
+            "select current_coin_price from scout_history order by rowid desc limit 1"
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:602]:
```suggestion
            "SELECT current_coin_price FROM scout_history ORDER BY rowid DESC LIMIT 1"
```

File: btb_manager_telegram/buttons.py [hunk 9]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+        )
+        lastCoinUSD = cur.fetchall()[0][0]
+
+        lastCoinFiatValue = lastCoinValue * lastCoinUSD
+
+        if lastCoinID != initialCoinID and initialCoinID != "":
+            cur.execute(
+                "select id from pairs where from_coin_id='{}' and to_coin_id='{}'".format(
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:610]:
```suggestion
                "SELECT id FROM pairs WHERE from_coin_id='{}' and to_coin_id='{}'".format(
```

File: btb_manager_telegram/buttons.py [hunk 10]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+                    lastCoinID, initialCoinID
+                )
+            )
+            pairID = cur.fetchall()[0][0]
+            cur.execute(
+                "select other_coin_price from scout_history where pair_id='{}' order by id desc limit 1".format(
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:616]:
```suggestion
                "SELECT other_coin_price FROM scout_history WHERE pair_id='{}' ORDER BY id DESC LIMIT 1".format(
```

File: btb_manager_telegram/buttons.py [hunk 11]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+                    pairID
+                )
+            )
+            currentValInitialCoin = cur.fetchall()[0][0]
+        else:
+            cur.execute(
+                "select current_coin_price from scout_history order by id desc limit 1"
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:623]:
```suggestion
                "SELECT current_coin_price FROM scout_history ORDER BY id DESC LIMIT 1"
```

File: btb_manager_telegram/buttons.py [hunk 12]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+            message += f"\nBot start coin is {firstTradeCoin} but currently not found in supported list."
+        elif initialCoinID == "":
+            message += "\nBot start coin not found in supported list."
+
+        message += "\n\n*Coin progress:*\n"
+        rows = []
+        # Compute Mini Coin Progress
+        for coin in coinList:
+            jumps = cur.execute(
+                f"select count(*) from trade_history where alt_coin_id='{coin[0]}' and selling=0 and state='COMPLETE'"
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:693]:
```suggestion
                f"SELECT COUNT(*) FROM trade_history WHERE alt_coin_id='{coin[0]}' and selling=0 and state='COMPLETE'"
```

File: btb_manager_telegram/buttons.py [hunk 13]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+            ).fetchall()[0][0]
+            if jumps > 0:
+                first_date = cur.execute(
+                    f"select datetime from trade_history where alt_coin_id='{coin[0]}' and selling=0 and state='COMPLETE' order by id asc limit 1"
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:697]:
```suggestion
                    f"SELECT datetime FROM trade_history WHERE alt_coin_id='{coin[0]}' and selling=0 and state='COMPLETE' ORDER BY id ASC LIMIT 1"
```

File: btb_manager_telegram/buttons.py [hunk 14]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+                ).fetchall()[0][0]
+                first_value = cur.execute(
+                    f"select alt_trade_amount from trade_history where alt_coin_id='{coin[0]}' and selling=0 and state='COMPLETE' order by id asc limit 1"
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:700]:
```suggestion
                    f"SELECT alt_trade_amount FROM trade_history WHERE alt_coin_id='{coin[0]}' and selling=0 and state='COMPLETE' ORDER BY id ASC LIMIT 1"
```

File: btb_manager_telegram/buttons.py [hunk 15]
```diff
@@ -523,6 +525,209 @@ def trade_history():
...
+                ).fetchall()[0][0]
+                last_value = cur.execute(
+                    f"select alt_trade_amount from trade_history where alt_coin_id='{coin[0]}' and selling=0 and state='COMPLETE' order by id desc limit 1"
```
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:703]:
```suggestion
                    f"SELECT alt_trade_amount FROM trade_history WHERE alt_coin_id='{coin[0]}' and selling=0 and state='COMPLETE' ORDER BY id DESC LIMIT 1"
```

File: btb_manager_telegram/buttons.py (hunk 10 above)
Comment (DmytroLitvinov) [inline btb_manager_telegram/buttons.py:616]:
```suggestion
                "SELECT other_coin_price FROM scout_history WHERE pair_id='{}' ORDER BY id DESC LIMIT 1".format(
```

**Review by DmytroLitvinov — CHANGES_REQUESTED** [review]

**Comment by titulebolide** [comment]
This bug as been pointed out:
Traceback (most recent call last):
  File "/home/ubuntu/binance-trade-bot/Binance1/binance-trade-bot/custom_scripts/bot_stats.py", line 20, in <module>
    initialCoinID = db.fetchall()[0][0]
IndexError: list index out of range

File: btb_manager_telegram/buttons.py [hunk 16]
```diff
@@ -703,20 +705,22 @@ def bot_stats():
...
+                rows.append(
                     [
                         coin[0],
-                        "{:.2f}".format(first_value),
-                        "{:.2f}".format(last_value),
-                        "{:.1f}".format(grow),
-                        "{}".format(jumps),
+                        float(first_value),
+                        float(last_value),
+                        float(grow),
```
Comment (ullasbharadwaj) [inline btb_manager_telegram/buttons.py:711]: 2 decimals is sufficient for growth percentage ? Looks neat.

**Comment by ullasbharadwaj** [comment]
You mean in profit calculation? I assumed, it makes sense to see the profit in USD as the current holding to the HODL holding which means profit percentage in terms of Coin/USD is the same. But if you see the profit in Fiat as the change between USDValueAtBotStart and CurrentUSDValue, then we will have different profit percentages.

**Comment by titulebolide** [comment]
Ohhh that's why! It was unclear, I thought the profit was relative to the start of the bot. IMO it is clearer to show the actual profit in USD (thus relative to the start) and the actual profit in the start coin (which then is equivalent to the profit relative to the HODL)

**Comment by ullasbharadwaj** [comment]
My bad, it was not clear. We can modify as you said, then we have 2 perspectives of visualising the profit. 👍

**Comment by titulebolide** [comment]
Here it is! 
`Profit       : -3.96% USD / +5.26% QTUM`

**Comment by titulebolide** [comment]
If you are happy with how it is, it is ready to be merged

**Comment by ullasbharadwaj** [comment]
Yes perfect. Thanks for great work.

**Pull request merged**